wallet: importdescriptors update existing

S3RK opened the pull request:
Rationale: allow updating existing descriptors with `importdescriptors` command.

Currently if you run same `importdescriptors` command twice with a descriptor containing private key you will get very confusing error — `Missing required fields`. What happens is that Wallet tries to write imported private key to the disk, but it exists already so we get `DB_KEYEXIST (-30995)` from BerkelyDB. Please note, that we set `DB_NOOVERWRITE` (I guess not to lose some keys accidentally). The exception is caught in `catch (...)` in rpcdump.cpp with a generic error.

With this PR if a descriptor is already present than we will update its activeness, internalness, label, range and next_index.
For the range only expansion is allowed (range start can only decrease, range end increase).

elichai commented:
Could we change `BerkeleyBatch::WriteKey` to return an enum describing the error? that will be inconsistent with the current API though.
but it will allow us to print a good error for the user, as he should probably know he already have that descriptor

achow101 commented:
I think we should go with option 2: To update the existing spkman. The keys should either already exist or are being imported at that time. I don't think it's good to add keys in memory and not write them to disk as this could potentially result in lost keys.

S3RK commented:
> Could we change `BerkeleyBatch::WriteKey` to return an enum describing the error? that will be inconsistent with the current API though.
> but it will allow us to print a good error for the user, as he should probably know he already have that descriptor

@elichai I believe it's better to process the command rather than just show better error message, because it allows to modify some meta data: for example descriptor's range. However I added a trivial commit to make error message less misleading.

> I think we should go with option 2: To update the existing spkman. The keys should either already exist or are being imported at that time. I don't think it's good to add keys in memory and not write them to disk as this could potentially result in lost keys.

@achow101 Indeed, this aligns with my intuition as well. I changed the implementation and it looks cleaner to me. Hope I got the update semantics correct for the `DescriptorScriptPubKeyMan`. Could you please check?

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -2254,3 +2254,22 @@ const std::vector<CScript> DescriptorScriptPubKeyMan::GetScriptPubKeys() const
+    if (!HasWalletDescriptor(descriptor)) {
+        throw std::runtime_error(std::string(__func__) + ": can only update matching descriptor");
+    }
+    if (descriptor.range_start > m_wallet_descriptor.range_start) {
```
We also need to check that new range_end >= old range_end.

achow101 reviewed (commented):
The commit adding the test should be after or be the same commit as the one adding the fix. This way all commits pass tests.

It would be nice if the tests checked the update behavior (e.g. new ranges, changed active-ness, etc.) but we can do that in a followup.

S3RK commented:
> The commit adding the test should be after or be the same commit as the one adding the fix. This way all commits pass tests.
> 
> It would be nice if the tests checked the update behavior (e.g. new ranges, changed active-ness, etc.) but we can do that in a followup.

@achow101 I squashed the test and the fix commits into one and extracted new function `CanUpdateToWalletDescriptor` to avoid code duplication. I also added some tests to verify various update scenarios: label change, range change, next_index, internal flag change, descriptor activation. A minor change were required in `DescriptorScriptPubKeyMan::SetCache` to make the tests pass.

Please note, descriptor deactivation is not working right now so there is no test for it. I'm currently working on it and will do as a separate PR.

S3RK commented:
Rebased and removed 51ea4c6ad808f881d26ff99da620e09d9549163e as it was already added to master

S3RK commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -2183,7 +2183,7 @@ void DescriptorScriptPubKeyMan::SetCache(const DescriptorCache& cache)
         std::vector<CScript> scripts_temp;
         if (!m_wallet_descriptor.descriptor->ExpandFromCache(i, m_wallet_descriptor.cache, scripts_temp, out_keys)) {
-            throw std::runtime_error("Error: Unable to expand wallet descriptor from cache");
+            continue;
```
Done. The cache is now regenerated on import.

S3RK commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -2254,3 +2254,41 @@ const std::vector<CScript> DescriptorScriptPubKeyMan::GetScriptPubKeys() const
     return script_pub_keys;
 }
+
+void DescriptorScriptPubKeyMan::UpdateWalletDescriptor(WalletDescriptor& descriptor)
```
Thanks. Changed the code to properly handle this case and added a test for it as well.

S3RK commented on src/wallet/wallet.cpp:
```diff
@@ -4478,10 +4478,15 @@ void CWallet::LoadActiveScriptPubKeyMan(uint256 id, OutputType type, bool intern
     spk_mans[type] = spk_man;
 
+    if (spk_mans_other[type] == spk_man) {
+        spk_mans_other[type] = nullptr;
```
I analyzed the code and couldn't find any uses of `LoadActiveScriptPubKeyMan` for legacy wallets. The function is only called when 1) we create new wallet, 2) call `importdescriptors` command or 3) when loading a wallet from disk. But if I'm not mistaken there is no active SPKs on disk for legacy wallets. That makes me believe this function is only relevant for descriptor wallets. So I put a guard at the top to throw an exception when we call it for legacy wallet. And all wallet functional tests pass on my laptop for now. I'm happy to reproduce any scenario in which this code should break legacy wallet.

achow101 replied:
Hmm, `SetupLegacyScriptPubKeyMan` adds the `LegacySPKM` to the maps that track internal/external, but doesn't write them to disk. I thought it wrote them to disk and now I'm wondering whether not doing so is a bug. But since it doesn't, that's why this works.

jonatack replied:
cead7ed ISTM this change should have test coverage (more so given the discussion) to spec the behavior and provide regression coverage. Edit: Ah, if I understand correctly some test coverage is added in the next commit b66b26a.

S3RK commented:
Added `wallet: deactivate descriptor` commit from #19774

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1836,6 +1836,10 @@ void DescriptorScriptPubKeyMan::MarkUnusedAddresses(const CScript& script)
 void DescriptorScriptPubKeyMan::AddDescriptorKey(const CKey& key, const CPubKey &pubkey)
 {
     LOCK(cs_desc_man);
+    if (m_map_keys.find(pubkey.GetID()) != m_map_keys.end()) {
```
I think it would be better to have this inside of `AddDescriptorKeyWithDB`. Additionally, it needs to check `m_map_crypted_keys` otherwise this won't work when the wallet is encrypted.

S3RK replied:
Done

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -4494,6 +4494,22 @@ void CWallet::LoadActiveScriptPubKeyMan(uint256 id, OutputType type, bool intern
+void CWallet::DeactivateScriptPubKeyMan(uint256 id, OutputType type, bool internal)
+{
+    auto& spk_mans = internal ? m_internal_spk_managers : m_external_spk_managers;
+    if (spk_mans[type] && spk_mans[type]->GetID() == id) {
```
I think using `[]` for access here could be problematic if the `type` isn't already in the map (I think that can happen). I would prefer using `find` here and then if it works, checking that the returned object is not `nullptr`.

S3RK replied:
I'm not a cpp expert, but the docs say in case when there is no such key a default value will be inserted. 
> std::map<Key,T,Compare,Allocator>::operator[]
> Returns a reference to the value that is mapped to a key equivalent to key, performing an insertion if such key does not already exist.
So I believe it safe in this particular case.

I still prefer `spk_mans[type]` over `spk_mans.find(type) != spk_mans.end() && spk_mans[type] != nullptr` but if that's going against the norms for the project I can change to whatever is considered standard way.

achow101 replied:
Right, but the point is that we don't want a default value inserted.

S3RK replied:
Done

S3RK commented:
@achow101 thanks for your time and review. I pushed changes as fixup commits for easier incremental review, when you believe the code is ready to be merged I'll squash them into original commits.

achow101 commented:
ACK, please squash.

achow101 commented:
ACK 908e18e79847155144111bafa2e6e5dbd5c6511c

achow101 reviewed (approved):
re-ACK e06c123bbcb3be5b7cea4a9480af48c8b54daa83

S3RK commented:
This should also fix #21716

achow101 commented:
re-ACK 5d9670462219b8c753516bfbdc5c7fa23046cd07

meshcollider commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -2280,3 +2286,38 @@ bool DescriptorScriptPubKeyMan::GetDescriptorString(std::string& out, bool priv)
+    m_wallet_descriptor = descriptor;
+}
+
+bool DescriptorScriptPubKeyMan::CanUpdateToWalletDescriptor(const WalletDescriptor& descriptor, std::string& error)
```
I feel like the name of this function is confusing and could be improved.

S3RK replied:
I've spent some time, but couldn't come up with anything better. Any suggestions?

meshcollider reviewed (commented):
Concept ACK + light code review, I'll finish reviewing soon

jonatack commented:
I'll try to review this later today.

jonatack commented on src/wallet/walletdb.cpp:
```diff
@@ -209,6 +209,12 @@ bool WalletBatch::WriteActiveScriptPubKeyMan(uint8_t type, const uint256& id, bo
 
+bool WalletBatch::EraseActiveScriptPubKeyMan(uint8_t type, bool internal)
+{
+    std::string key = internal ? DBKeys::ACTIVEINTERNALSPK : DBKeys::ACTIVEEXTERNALSPK;
```
5d96704 style nit
```suggestion
    const std::string key{internal ? DBKeys::ACTIVEINTERNALSPK : DBKeys::ACTIVEEXTERNALSPK};
```

S3RK replied:
done

jonatack commented on src/wallet/wallet.cpp:
```diff
@@ -4625,12 +4625,38 @@ void CWallet::AddActiveScriptPubKeyMan(uint256 id, OutputType type, bool interna
+        WalletLogPrintf("Deactivate spkMan: id = %s, type = %d, internal = %d\n", id.ToString(), static_cast<int>(type), static_cast<int>(internal));
+        WalletBatch batch(GetDatabase());
+        if (!batch.EraseActiveScriptPubKeyMan(static_cast<uint8_t>(type), internal)) {
+            throw std::runtime_error(std::string(__func__) + ": erasing active ScriptPubKeyMan id failed");
```
5d96704 test coverage?

S3RK replied:
Idk how to cover this. One need to introduce a low-level database error to hit this line. I'm not sure it's worth it

jonatack commented on src/wallet/rpcdump.cpp:
```diff
@@ -1564,16 +1563,19 @@ static UniValue ProcessDescriptorImport(CWallet& wallet, const UniValue& data, c
+        } else {
+            if (w_desc.descriptor->GetOutputType()) {
+                wallet.DeactivateScriptPubKeyMan(spk_manager->GetID(), *w_desc.descriptor->GetOutputType(), internal);
+            }
```
5d9670462219b8c753516bfbdc5c7fa23046cd07 does this have test coverage?

S3RK replied:
This is covered in the section "Check can deactivate active descriptor". You can try deleting this lines and the test will fail

jonatack reviewed (commented):
Concept/Approach ACK 5d9670462219b8c753516bfbdc5c7fa23046cd07 reviewed, debug-built and ran unit tests and wallet_importdescriptors.py on each commit. I don't know this part of the codebase well; the code and behavioral changes seem ok but my concern is if there is test coverage of all the changes in behavior.

fjahr commented on src/wallet/rpcdump.cpp:
```diff
@@ -1546,7 +1546,7 @@ static UniValue ProcessDescriptorImport(CWallet& wallet, const UniValue& data, c
         if (existing_spk_manager) {
             if (!existing_spk_manager->CanUpdateToWalletDescriptor(w_desc, error)) {
-                throw JSONRPCError(RPC_INVALID_PARAMS, error);
+                throw JSONRPCError(RPC_INVALID_PARAMETER, error);
```
This could just be fixed in the previous commit where this line is added

S3RK replied:
Done

fjahr commented on src/wallet/wallet.cpp:
```diff
@@ -4625,12 +4625,38 @@ void CWallet::AddActiveScriptPubKeyMan(uint256 id, OutputType type, bool interna
 void CWallet::LoadActiveScriptPubKeyMan(uint256 id, OutputType type, bool internal)
 {
+    if (!IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS)) {
+        throw std::runtime_error("Cannot LoadActiveScriptPubKeyMan to a non-descriptor wallet\n");
```
Is the newline needed here? Doesn't seem to be the standard.

S3RK replied:
This is just a guard and should be never hit unless there is a logical error in the code.
Replaced with Assert and added a comment

fjahr commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -2280,3 +2286,38 @@ bool DescriptorScriptPubKeyMan::GetDescriptorString(std::string& out, bool priv)
+        return false;
+    }
+    // Use inclusive range for error
+    auto range = strprintf("current range = [%d,%d]", m_wallet_descriptor.range_start, m_wallet_descriptor.range_end-1);
```
nit: I prefer to not create helper variables outside of error handling that are only used inside of error handling so I would inline this. I also think this code be reorganized slightly to make the error cumulative message, i.e. if both errors are hit the error message can include both messages about `range_start` and `range_end`.

S3RK replied:
Restructured the code

fjahr reviewed (commented):
Concept ACK

My comments are not that important but I agree with @meshcollider that `CanUpdateToWalletDescriptor` could be named better and with @jonatack that test coverage should be improved.

S3RK committed:
wallet: allow to import same descriptor twice

S3RK committed:
wallet: don't mute exceptions in importdescriptors

S3RK committed:
wallet: maintain SPK consistency on internal flag change

S3RK committed:
test: wallet importdescriptors update existing

S3RK committed:
wallet: deactivate descriptor

S3RK commented on src/wallet/rpcdump.cpp:
```diff
@@ -1564,16 +1563,19 @@ static UniValue ProcessDescriptorImport(CWallet& wallet, const UniValue& data, c
         result.pushKV("success", UniValue(false));
-
-        result.pushKV("error", JSONRPCError(RPC_MISC_ERROR, "Missing required fields"));
+        result.pushKV("error", JSONRPCError(RPC_MISC_ERROR, e.what()));
```
Hm... thanks for pointing to that. Now looked back at that piece, I decided it's better to just drop it.

achow101 commented:
re-ACK 3efaf83c75cd8dc2fa084537b8ed6715fb58c04d

meshcollider reviewed (commented):
Code review ACK 3efaf83c75cd8dc2fa084537b8ed6715fb58c04d

jonatack commented on test/functional/wallet_importdescriptors.py:
```diff
@@ -252,6 +255,39 @@ def run_test(self):
+        self.test_importdesc({**range_request, "range": [0, 10]}, wallet=wpriv, success=False,
+                             error_code=-8, error_message='new range must include current range = [0,20]')
+        self.test_importdesc({**range_request, "range": [5, 20]}, wallet=wpriv, success=False,
+                             error_code=-8, error_message='new range must include current range = [0,20]')
```
6737d965 suggestion, separate the salient from the redundant
```diff
-        self.test_importdesc({**range_request, "range": [5, 10]}, wallet=wpriv, success=False,
-                             error_code=-8, error_message='new range must include current range = [0,20]')
-        self.test_importdesc({**range_request, "range": [0, 10]}, wallet=wpriv, success=False,
-                             error_code=-8, error_message='new range must include current range = [0,20]')
-        self.test_importdesc({**range_request, "range": [5, 20]}, wallet=wpriv, success=False,
-                             error_code=-8, error_message='new range must include current range = [0,20]')
+        for r in [[5, 10], [0, 10], [5, 20]]:
+            self.test_importdesc(req={**range_request, "range": r}, wallet=wpriv, success=False, error_code=-8, error_message="new range must include current range = [0,20]")
```

jonatack commented on test/functional/wallet_importdescriptors.py:
```diff
@@ -89,22 +88,26 @@ def run_test(self):
-                             success=True)
+        import_request = {"desc": descsum_create("pkh(" + key.pubkey + ")"),
+                 "timestamp": "now",
+                 "label": "Descriptor import test"}
```
bf68ebc formatting nit

```diff
         import_request = {"desc": descsum_create("pkh(" + key.pubkey + ")"),
-                 "timestamp": "now",
-                 "label": "Descriptor import test"}
+                          "timestamp": "now",
+                          "label": "Descriptor import test"}
```

or

```diff
-        import_request = {"desc": descsum_create("pkh(" + key.pubkey + ")"),
-                 "timestamp": "now",
-                 "label": "Descriptor import test"}
+        import_request = {
+            "desc": descsum_create("pkh(" + key.pubkey + ")"),
+            "timestamp": "now",
+            "label": "Descriptor import test",
+        }
```

jonatack commented on test/functional/wallet_importdescriptors.py:
```diff
@@ -252,6 +255,39 @@ def run_test(self):
+        assert_equal(wpriv.getwalletinfo()['keypoolsize'], 21)
+        # Can keep range the same
+        self.test_importdesc({**range_request, "range": [0, 20]}, wallet=wpriv, success=True)
+        assert_equal(wpriv.getwalletinfo()['keypoolsize'], 21)
```
6737d96 readability nit
```diff
         self.log.info("Verify we can only extend descriptor's range")
         range_request = {"desc": descsum_create(desc), "timestamp": "now", "range": [5, 10], 'active': True}
+
         self.test_importdesc(range_request, wallet=wpriv, success=True)
         assert_equal(wpriv.getwalletinfo()['keypoolsize'], 6)
+
         self.test_importdesc({**range_request, "range": [0, 10]}, wallet=wpriv, success=True)
         assert_equal(wpriv.getwalletinfo()['keypoolsize'], 11)
+
         self.test_importdesc({**range_request, "range": [0, 20]}, wallet=wpriv, success=True)
         assert_equal(wpriv.getwalletinfo()['keypoolsize'], 21)
+
         # Can keep range the same
         self.test_importdesc({**range_request, "range": [0, 20]}, wallet=wpriv, success=True)
         assert_equal(wpriv.getwalletinfo()['keypoolsize'], 21)
```

jonatack reviewed (commented):
Light ACK 3efaf83c75cd8dc2fa084537b8ed6715fb58c04d per `git range-diff a000cb0 5d96704 3efaf83` and as a sanity check, re-debug-built on debian with gcc 10.2.1 and clang 11, ran wallet_importdescriptors.py

Some touch-ups if you update.
